File: native/camera/CameraRenderer.swift
// SwiftDotNet.Controls.Camera — Apple (AVFoundation + Vision) renderer.
//
// This is the native half of the `CameraView` control on iOS (and Mac Catalyst). It is NOT part of the C#
// NuGet package (custom renderers on Apple are Swift, registered via `swiftDotNetRegisterRenderer`). Ship it
// by adding this file to the app's Swift bridge sources (or a companion Swift package linked into the app)
// and calling `registerSwiftDotNetCameraRenderer()` at startup, alongside the SwiftDotNetBridge — exactly
// like native/maps/MapRenderer.swift.
//
// It reads the props the C# `CameraView` writes ("facing", "flash", "analyzers", "captureToken"), hosts a
// live AVCaptureVideoPreviewLayer, runs Vision requests (barcode/face/text) on the video frames, and emits
// results back over the single event channel with the `kind:body` grammar CameraView defines
// ("photo:<b64>", "barcode:<fmt>:<b64>", "faces:<n>", "text:<b64>", "focus:<x>,<y>", "error:<b64>").
//
// STATUS: authored for review — not compiled in this environment (needs the Apple SDK + the bridge build).
// The AVFoundation/Vision usage is type-checked standalone against the iOS SDK via a bridge-symbol stub.

#if canImport(UIKit) && canImport(AVFoundation)
import SwiftUI
import AVFoundation
import Vision
import SwiftDotNetBridge   // companion module: swiftDotNetRegisterRenderer + SwiftDotNetProps + emitEvent

// MARK: - The capture controller

@MainActor
final class CameraController: NSObject, ObservableObject,
    AVCaptureVideoDataOutputSampleBufferDelegate, AVCapturePhotoCaptureDelegate {

    // AVCaptureSession isn't Sendable; the unsafe opt-out lets the capture queue start/stop it (the standard
    // AVFoundation pattern — the session is only mutated on the main actor or the dedicated queue).
    nonisolated(unsafe) let session = AVCaptureSession()
    private let photoOutput = AVCapturePhotoOutput()
    private let videoOutput = AVCaptureVideoDataOutput()
    private let queue = DispatchQueue(label: "swiftdotnet.camera")
    private var deviceInput: AVCaptureDeviceInput?

    var nodeId: String = ""
    var analyzers: Set<String> = []
    private var lastFaceCount = -1
    private var lastText = ""
    private var busy = false

    func configure(facing: String, analyzers: Set<String>) {
        self.analyzers = analyzers
        session.beginConfiguration()
        session.sessionPreset = .high

        // (Re)bind the camera input for the requested facing.
        if let existing = deviceInput { session.removeInput(existing) }
        let position: AVCaptureDevice.Position = facing == "front" ? .front : .back
        if let device = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: position),
           let input = try? AVCaptureDeviceInput(device: device),
           session.canAddInput(input) {
            session.addInput(input)
            deviceInput = input
        } else {
            emitEvent(nodeId, "error:" + b64("Camera unavailable"))
        }

        if session.canAddOutput(photoOutput) { session.addOutput(photoOutput) }
        if !analyzers.isEmpty, session.canAddOutput(videoOutput) {
            videoOutput.setSampleBufferDelegate(self, queue: queue)
            session.addOutput(videoOutput)
        }
        session.commitConfiguration()
        if !session.isRunning { queue.async { self.session.startRunning() } }
    }

    func capture() {
        let settings = AVCapturePhotoSettings()
        photoOutput.capturePhoto(with: settings, delegate: self)
    }

    // Photo captured → emit the JPEG bytes.
    nonisolated func photoOutput(_ output: AVCapturePhotoOutput,
                                 didFinishProcessingPhoto photo: AVCapturePhoto, error: Error?) {
        guard let data = photo.fileDataRepresentation() else { return }
        Task { @MainActor in emitEvent(self.nodeId, "photo:" + data.base64EncodedString()) }
    }

    // Each video frame → run the enabled Vision requests.
    nonisolated func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer,
                                   from connection: AVCaptureConnection) {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        Task { @MainActor in self.analyze(pixelBuffer) }
    }

    private func analyze(_ pixelBuffer: CVPixelBuffer) {
        if busy { return }
        busy = true
        let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, options: [:])
        var requests: [VNRequest] = []

        if analyzers.contains("barcode") {
            requests.append(VNDetectBarcodesRequest { [weak self] req, _ in
                guard let self, let r = (req.results as? [VNBarcodeObservation])?.first,
                      let payload = r.payloadStringValue else { return }
                let fmt = r.symbology.rawValue
                Task { @MainActor in emitEvent(self.nodeId, "barcode:\(fmt):" + self.b64(payload)) }
            })
        }
        if analyzers.contains("face") {
            requests.append(VNDetectFaceRectanglesRequest { [weak self] req, _ in
                guard let self else { return }
                let count = (req.results as? [VNFaceObservation])?.count ?? 0
                if count != self.lastFaceCount {
                    self.lastFaceCount = count
                    Task { @MainActor in emitEvent(self.nodeId, "faces:\(count)") }
                }
            })
        }
        if analyzers.contains("text") {
            let textReq = VNRecognizeTextRequest { [weak self] req, _ in
                guard let self else { return }
                let text = (req.results as? [VNRecognizedTextObservation])?
                    .compactMap { $0.topCandidates(1).first?.string }
                    .joined(separator: "\n") ?? ""
                if !text.isEmpty, text != self.lastText {
                    self.lastText = text
                    Task { @MainActor in emitEvent(self.nodeId, "text:" + self.b64(text)) }
                }
            }
            textReq.recognitionLevel = .fast
            requests.append(textReq)
        }

        try? handler.perform(requests)
        busy = false
    }

    private nonisolated func b64(_ s: String) -> String {
        Data(s.utf8).base64EncodedString()
    }
}

// MARK: - The preview view (UIViewRepresentable wrapping the AVCaptureVideoPreviewLayer)

final class PreviewUIView: UIView {
    override class var layerClass: AnyClass { AVCaptureVideoPreviewLayer.self }
    var previewLayer: AVCaptureVideoPreviewLayer { layer as! AVCaptureVideoPreviewLayer }
}

struct CameraPreview: UIViewRepresentable {
    let controller: CameraController
    let onTapFocus: (CGPoint) -> Void

    func makeUIView(context: Context) -> PreviewUIView {
        let view = PreviewUIView()
        view.previewLayer.session = controller.session
        view.previewLayer.videoGravity = .resizeAspectFill
        let tap = UITapGestureRecognizer(target: context.coordinator, action: #selector(Coordinator.handleTap(_:)))
        view.addGestureRecognizer(tap)
        context.coordinator.view = view
        return view
    }

    func updateUIView(_ uiView: PreviewUIView, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(onTapFocus) }

    final class Coordinator: NSObject {
        weak var view: PreviewUIView?
        let onTapFocus: (CGPoint) -> Void
        init(_ onTapFocus: @escaping (CGPoint) -> Void) { self.onTapFocus = onTapFocus }

        @objc func handleTap(_ g: UITapGestureRecognizer) {
            guard let view else { return }
            let p = g.location(in: view)
            // Normalize to 0–1 within the preview bounds.
            onTapFocus(CGPoint(x: p.x / max(view.bounds.width, 1), y: p.y / max(view.bounds.height, 1)))
        }
    }
}

// MARK: - The SwiftDotNet-facing view

struct SwiftDotNetCameraView: View {
    let props: SwiftDotNetProps
    @StateObject private var controller = CameraController()

    private var facing: String { props.string("facing") ?? "back" }
    private var analyzerSet: Set<String> {
        Set((props.string("analyzers") ?? "").split(separator: ",").map(String.init))
    }
    private var captureToken: Double { props.number("captureToken") ?? 0 }

    var body: some View {
        CameraPreview(controller: controller) { pt in
            emitEvent(props.id, "focus:\(pt.x),\(pt.y)")
        }
        .onAppear {
            controller.nodeId = props.id
            controller.configure(facing: facing, analyzers: analyzerSet)
        }
        .onChange(of: facing) { _, f in controller.configure(facing: f, analyzers: analyzerSet) }
        .onChange(of: captureToken) { _, _ in controller.capture() }
    }
}

// MARK: - Registration

/// Call once at app startup (after loading the SwiftDotNetBridge) to render `CameraView` nodes via AVFoundation.
public func registerSwiftDotNetCameraRenderer() {
    swiftDotNetRegisterRenderer("CameraView") { props in AnyView(SwiftDotNetCameraView(props: props)) }
}

/// C entry point so managed code can register the camera renderer at startup (resolved via dlsym / "__Internal").
@_cdecl("swiftdotnet_register_camera")
public func swiftdotnet_register_camera() {
    registerSwiftDotNetCameraRenderer()
}

#endif

File: native/camera/CameraRenderer.kt
// SwiftDotNet.Controls.Camera — Android (CameraX + ML Kit) renderer.
//
// The native half of the `CameraView` control on Android. Custom renderers on Android are Kotlin,
// registered via `registerRenderer`, so this is NOT part of the C# NuGet package. Ship it by adding this
// file to the SwiftDotNetComposeBridge sources (rebuilding the AAR) or a companion module the app links,
// and calling `registerSwiftDotNetCameraRenderer()` at startup — exactly like native/maps/MapRenderer.kt.
//
// It reads the props the C# `CameraView` writes ("facing", "flash", "analyzers", "captureToken"), hosts a
// CameraX PreviewView in an AndroidView, runs ML Kit frame analysis, and emits results back over the single
// event channel with the `kind:body` grammar CameraView defines ("photo:<b64>", "barcode:<fmt>:<b64>",
// "faces:<n>", "text:<b64>", "focus:<x>,<y>", "error:<b64>"). Mirrors native/camera/CameraRenderer.swift.
//
// Requires CameraX + ML Kit:
//   implementation("androidx.camera:camera-camera2:1.3.+")
//   implementation("androidx.camera:camera-lifecycle:1.3.+")
//   implementation("androidx.camera:camera-view:1.3.+")
//   implementation("com.google.mlkit:barcode-scanning:17.+")
//   implementation("com.google.mlkit:face-detection:16.+")
//   implementation("com.google.mlkit:text-recognition:16.+")
//
// STATUS: authored for review — not compiled in this environment (needs the CameraX/ML Kit deps + AAR rebuild).

package com.swiftdotnet.bridge

import android.util.Base64
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.ImageProxy
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalLifecycleOwner
import androidx.compose.ui.viewinterop.AndroidView
import androidx.core.content.ContextCompat
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.io.ByteArrayOutputStream
import java.util.concurrent.Executors

private fun b64(s: String) = Base64.encodeToString(s.toByteArray(Charsets.UTF_8), Base64.NO_WRAP)

@Composable
fun SwiftDotNetCamera(props: SwiftDotNetProps) {
    val context = LocalContext.current
    val lifecycleOwner = LocalLifecycleOwner.current
    val previewView = remember { PreviewView(context) }
    val executor = remember { Executors.newSingleThreadExecutor() }
    val imageCapture = remember { ImageCapture.Builder().build() }

    val facing = props.string("facing") ?: "back"
    val analyzers = (props.string("analyzers") ?: "").split(",").filter { it.isNotEmpty() }.toSet()
    val captureToken = props.string("captureToken") ?: ""

    // ML Kit detectors (created once for the enabled analyzers).
    val barcodeScanner = remember { BarcodeScanning.getClient() }
    val faceDetector = remember { FaceDetection.getClient() }
    val textRecognizer = remember { TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS) }
    var lastFaceCount = remember { -1 }
    var lastText = remember { "" }

    // Bind the camera + analysis to the lifecycle whenever the facing changes.
    LaunchedEffect(facing) {
        val provider = ProcessCameraProvider.getInstance(context).get()
        val selector = if (facing == "front") CameraSelector.DEFAULT_FRONT_CAMERA else CameraSelector.DEFAULT_BACK_CAMERA
        val preview = Preview.Builder().build().also { it.setSurfaceProvider(previewView.surfaceProvider) }

        val analysis = ImageAnalysis.Builder()
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .build()
        if (analyzers.isNotEmpty()) {
            analysis.setAnalyzer(executor) { proxy: ImageProxy ->
                val media = proxy.image
                if (media == null) { proxy.close(); return@setAnalyzer }
                val input = InputImage.fromMediaImage(media, proxy.imageInfo.rotationDegrees)
                var pending = 0
                fun done() { if (--pending <= 0) proxy.close() }

                if (analyzers.contains("barcode")) {
                    pending++
                    barcodeScanner.process(input).addOnSuccessListener { list ->
                        list.firstOrNull()?.rawValue?.let { v ->
                            props.emit("barcode:${list[0].format}:${b64(v)}")
                        }
                    }.addOnCompleteListener { done() }
                }
                if (analyzers.contains("face")) {
                    pending++
                    faceDetector.process(input).addOnSuccessListener { faces ->
                        if (faces.size != lastFaceCount) { lastFaceCount = faces.size; props.emit("faces:${faces.size}") }
                    }.addOnCompleteListener { done() }
                }
                if (analyzers.contains("text")) {
                    pending++
                    textRecognizer.process(input).addOnSuccessListener { result ->
                        val t = result.text
                        if (t.isNotEmpty() && t != lastText) { lastText = t; props.emit("text:${b64(t)}") }
                    }.addOnCompleteListener { done() }
                }
                if (pending == 0) proxy.close()
            }
        }

        try {
            provider.unbindAll()
            provider.bindToLifecycle(lifecycleOwner, selector, preview, imageCapture, analysis)
        } catch (e: Exception) {
            props.emit("error:${b64(e.message ?: "Camera unavailable")}")
        }
    }

    // Capture a still whenever the C# captureToken changes.
    LaunchedEffect(captureToken) {
        if (captureToken.isEmpty() || captureToken == "0") return@LaunchedEffect
        imageCapture.takePicture(executor, object : ImageCapture.OnImageCapturedCallback() {
            override fun onCaptureSuccess(image: ImageProxy) {
                val buffer = image.planes[0].buffer
                val bytes = ByteArray(buffer.remaining()).also { buffer.get(it) }
                props.emit("photo:${Base64.encodeToString(bytes, Base64.NO_WRAP)}")
                image.close()
            }
            override fun onError(exc: ImageCaptureException) {
                props.emit("error:${b64(exc.message ?: "Capture failed")}")
            }
        })
    }

    AndroidView(
        factory = {
            previewView.setOnTouchListener { v, ev ->
                if (ev.action == android.view.MotionEvent.ACTION_UP) {
                    props.emit("focus:${ev.x / v.width.coerceAtLeast(1)},${ev.y / v.height.coerceAtLeast(1)}")
                }
                v.performClick(); true
            }
            previewView
        },
        modifier = Modifier.fillMaxSize(),
    )
}

/** Call once at app startup (after the bridge is initialized) to render `CameraView` nodes via CameraX. */
fun registerSwiftDotNetCameraRenderer() {
    registerRenderer("CameraView") { props -> SwiftDotNetCamera(props) }
}
